eva/storage/sqlite_storage_engine.py

```diff
@@ -0,0 +1,148 @@
+# coding=utf-8
+# Copyright 2018-2022 EVA
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Iterator, List
+
+import numpy as np
+import pandas as pd
+from sqlalchemy.ext.declarative import declarative_base
+
+from eva.catalog.column_type import ColumnType
+from eva.catalog.models.base_model import BaseModel
+from eva.catalog.models.df_column import DataFrameColumn
+from eva.catalog.models.df_metadata import DataFrameMetadata
+from eva.catalog.schema_utils import SchemaUtils
+from eva.catalog.sql_config import SQLConfig
+from eva.models.storage.batch import Batch
+from eva.storage.abstract_storage_engine import AbstractStorageEngine
+from eva.utils.generic_utils import PickleSerializer, get_size
+from eva.utils.logging_manager import logger
+
+# Leveraging Dynamic schema in SQLAlchemy
+# https://sparrigan.github.io/sql/sqla/2016/01/03/dynamic-tables.html
+
+Base = declarative_base()
+
+
+class SQLStorageEngine(AbstractStorageEngine):
+    def __init__(self):
+        """
+        Grab the existing sql session
+        """
+        self._sql_session = SQLConfig().session
+        self._sql_engine = SQLConfig().engine
+        self._serializer = PickleSerializer()
+
+    def _dict_to_sql_row(self, dict_row: dict, columns: List[DataFrameColumn]):
+        # Serialize numpy data
+        for col in columns:
+            if col.type == ColumnType.NDARRAY:
+                dict_row[col.name] = self._serializer.serialize(dict_row[col.name])
+            elif isinstance(dict_row[col.name], (np.generic,)):
+                # SqlAlchemy does not consume numpy generic data types
+                # convert numpy datatype to python generic datatype
+                # eg. np.int64 -> int
+                dict_row[col.name] = dict_row[col.name].tolist()
+        return dict_row
+
+    def _sql_row_to_dict(self, sql_row: tuple, columns: List[DataFrameColumn]):
+        # Deserialize numpy data
+        dict_row = {}
+        for idx, col in enumerate(columns):
+            if col.type == ColumnType.NDARRAY:
+                dict_row[col.name] = self._serializer.deserialize(sql_row[idx])
+            else:
+                dict_row[col.name] = sql_row[idx]
+        return dict_row
+
+    def create(self, table: DataFrameMetadata, **kwargs):
```

Comment:
What happen if the table is already existed? This raised an issue that I did not notice before. I was using `if_not_exists` in the [load executor]( https://github.com/georgia-tech-db/eva/blob/b8cfc51259cadc68b184b4ae2ef6da077cdcde6f/eva/executor/load_video_executor.py#L59). In [mat executor](https://github.com/georgia-tech-db/eva/blob/b8cfc51259cadc68b184b4ae2ef6da077cdcde6f/eva/executor/create_mat_view_executor.py#L34), we can use `handle_if_not_exists` because the catalog entry is created in the mat executor. But this is not the case for the load operator, the catalog entry is created before the load executor. I think we shall choose one way for both of them. And update the mat, load design and this `create` (e.g., doc what happens the table exists, remove `**kwargs`, add a call to check whether table exists or can we use read to do it?)

Reply:
Can you raise it as an issue? We can fix it later as I want to get rid of Pyspark asap